blog/tidy-market-microstructure/index.qmd

```diff
@@ -0,0 +1,1674 @@
+
+```{r}
+tick_size <- 0.5
+```
+
+Another dimension of quoted liquidity is the market depth. We measure the average depth quoted at the best bid and ask prices. It is defined as $quoted\_depth = (Q^A + Q^B)/2$, where $Q^A$ and $Q^B$ are the depths available at the bid and ask prices. 
+
+In the code below, we store the liquidity measures in a new `data.table` named `quotes_liquidity`. This is because the new variables are averages, observed on a ticker-date frequency, as opposed to the tick-by-tick frequency of the `quotes` object. We multiply the quoted spread by 10,000 to express it in basis points, and divide the quoted depth by 100,000 to express it in thousand GBP.
+
+The output shows that the liquidity is higher at the LSE than at the TQE. Both in nominal and relative terms, the spreads are somewhat tighter at the LSE, and there is more than three times more depth posted at the LSE.
+
+::: {.panel-tabset group="language"}
+## data.table
+```{r Ex ante liquidity}
+# Measure the average quote-based liquidity
+# This step calculates the mean of each of the market quality variables, for each 
+# ticker-day (as indicated in `by = c("ticker", "date")`)
+
+quotes_liquidity <- quotes[, {
+	quoted_spread = ask_price - bid_price
+	list(quoted_spread_nom = mean(quoted_spread, na.rm = TRUE),
+	     quoted_spread_relative = mean(quoted_spread / midpoint, na.rm = TRUE),
+	     quoted_spread_tick = mean(quoted_spread / tick_size, na.rm = TRUE),
+	     quoted_depth = mean(bid_depth * bid_price + ask_depth * ask_price, 
+	                         na.rm = TRUE) / 2)},
+	by = c("ticker", "date")]
+
+# Output the liquidity measures, averaged across the five trading days for each ticker. 
+quotes_liquidity[, 
+	list(quoted_spread_nom = round(mean(quoted_spread_nom), digits = 2),
+	     quoted_spread_relative = round(mean(quoted_spread_relative) * 1e4, digits = 2),
+	     quoted_spread_tick = round(mean(quoted_spread_tick), digits = 2),
+	     quoted_depth = round(mean(quoted_depth) * 1e-5, digits = 2)), 
+	by = "ticker"]
+```
+## tidyverse (using dtplyr)
+```{r}
+tv_quotes_liquidity <- tv_quotes |>
+  mutate(quoted_spread = ask_price - bid_price) |>
+  group_by(ticker, date) |> 
+  summarise(quoted_spread_nom = mean(quoted_spread, na.rm = TRUE),
+            quoted_spread_relative = mean(quoted_spread / midpoint, na.rm = TRUE) * 1e4,
+            quoted_spread_tick = mean(quoted_spread / tick_size, na.rm = TRUE),
+            quoted_depth = mean(bid_depth * bid_price + ask_depth * ask_price, 
+                                        na.rm = TRUE) / 2 * 1e-5)
+
+tv_quotes_liquidity |> 
+  group_by(ticker) |>
+  summarise(across(contains("quoted"), 
+                   ~round(mean(.), digits = 2))) |>
+  pivot_longer(-ticker) |>
+  as_tibble() |>
```

Comment:
as_tibble() strictly speaking not required or can at least be moved to end of pipe